cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/purgatory/Purgatory.java

```diff
@@ -0,0 +1,194 @@
+                                          PURGATORY_CLEANER_INITIAL_DELAY_SECONDS,
+                                          PURGATORY_CLEANER_PERIOD_SECONDS,
+                                          TimeUnit.SECONDS);
+  }
+
+  /**
+   * Add request to the purgatory and return the request info that has been added to the purgatory by its id.
+   *
+   * @param request Http Servlet Request to add to the purgatory.
+   * @param parameters Request parameters.
+   * @param <P> Type corresponding to the request parameters.
+   * @return The result showing the request info that has been added to the purgatory by its id.
+   */
+  public synchronized <P extends CruiseControlParameters> PurgatoryOrReviewResult addRequest(HttpServletRequest request,
+                                                                                             P parameters) {
+    if (!request.getMethod().equals("POST")) {
+      throw new IllegalArgumentException(String.format("Purgatory can only contain POST request (Attempted to add: %s).",
+                                                       httpServletRequestToString(request)));
+    }
+    RequestInfo requestInfo = new RequestInfo(request, parameters);
+    _requestInfoById.put(_requestId, requestInfo);
+
+    return new PurgatoryOrReviewResult(Collections.singletonMap(_requestId++, requestInfo));
+  }
+
+  /**
+   * Ensure that:
+   * <ul>
+   *   <li>A request with the given review id exists in the purgatory.</li>
+   *   <li>The request with the given review id matches the given request.</li>
+   *   <li>The request with the given review id is approved in the purgatory.</li>
+   * </ul>
+   *
+   * Then mark the review status as submitted.
+   *
+   * @param reviewId The review id for which the corresponding request is requested to be submitted.
+   * @param request The request to submit.
+   * @return Submitted request info.
+   */
+  public synchronized RequestInfo submit(int reviewId, HttpServletRequest request) {
+    RequestInfo requestInfo = _requestInfoById.get(reviewId);
+    // 1. Ensure that a request with the given review id exists in the purgatory.
+    if (requestInfo == null) {
+      throw new UserRequestException(
+          String.format("No request with review id %d exists in purgatory. Please use %s endpoint to check for the "
+                        + "current requests awaiting review in purgatory.", reviewId, REVIEW));
+    }
+
+    // 2. Ensure that the request with the given review id matches the given request.
+    EndPoint endpoint = ParameterUtils.endPoint(request);
+    if (requestInfo.endPoint() != endpoint) {
+      throw new UserRequestException(
+          String.format("Request with review id %d is associated with %s endpoint, but the given request has %s endpoint."
+                        + "Please use %s endpoint to check for the current requests awaiting review in purgatory.",
+                        reviewId, requestInfo.endPoint(), endpoint, REVIEW));
+    }
+
```

Comment:
Where do we check that the request has been approved?

Reply:
The response of `REVIEW` endpoint provides the approved requests.